token/sdk/vault/provider.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package vault
+
+import (
+	"sync"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view"
+	"github.com/hyperledger-labs/fabric-token-sdk/token"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/auditdb"
+	vault2 "github.com/hyperledger-labs/fabric-token-sdk/token/services/network/common/vault"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/tokendb"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/ttxdb"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/vault"
+	"github.com/pkg/errors"
+)
+
+type Provider struct {
+	sp view.ServiceProvider
+
+	vaultCacheLock sync.RWMutex
+	vaultCache     map[string]vault.Vault
+}
+
+func NewVaultProvider(sp view.ServiceProvider) *Provider {
+	return &Provider{sp: sp, vaultCache: make(map[string]vault.Vault)}
+}
+
+func (v *Provider) Vault(network string, channel string, namespace string) (vault.Vault, error) {
+	k := network + channel + namespace
+	// Check cache
+	v.vaultCacheLock.RLock()
+	res, ok := v.vaultCache[k]
+	v.vaultCacheLock.RUnlock()
+	if ok {
+		return res, nil
+	}
+
+	// lock
+	v.vaultCacheLock.Lock()
+	defer v.vaultCacheLock.Unlock()
+
+	// check cache again
+	res, ok = v.vaultCache[k]
+	if ok {
+		return res, nil
+	}
+
+	tmsID := token.TMSID{
+		Network:   network,
+		Channel:   channel,
+		Namespace: namespace,
+	}
+	tokenDB, err := tokendb.GetByTMSId(v.sp, tmsID)
```

Comment:
Let's avoid using SP and pass the token db provider instead. This is injected directly into the constructor.

Reply:
okay